// lateration/lib/vector.js
"use strict";

var _createClass = require("babel-runtime/helpers/create-class")["default"];

var _classCallCheck = require("babel-runtime/helpers/class-call-check")["default"];

var _interopRequireDefault = require("babel-runtime/helpers/interop-require-default")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _invariant = require("invariant");

var _invariant2 = _interopRequireDefault(_invariant);

var _utilsIsNumber = require("./utils/isNumber");

var _utilsIsNumber2 = _interopRequireDefault(_utilsIsNumber);

var Vector = (function () {
  function Vector() {
    var x = arguments.length <= 0 || arguments[0] === undefined ? 0 : arguments[0];
    var y = arguments.length <= 1 || arguments[1] === undefined ? 0 : arguments[1];

    _classCallCheck(this, Vector);

    (0, _invariant2["default"])(typeof x === "number", "Expected `x` to be a number.");
    (0, _invariant2["default"])(typeof y === "number", "Expected `y` to be a number.");

    (0, _invariant2["default"])((0, _utilsIsNumber2["default"])(x) && (0, _utilsIsNumber2["default"])(y), "Expected `x` and `y` to be numbers.");

    this.x = x;
    this.y = y;
  }

  _createClass(Vector, [{
    key: "normalize",
    value: function normalize() {
      var n = arguments.length <= 0 || arguments[0] === undefined ? 1 : arguments[0];

      (0, _invariant2["default"])(n !== 0, "NaN Error! Can't divide by 0.");

      this.x = this.x / n;
      this.y = this.y / n;
    }
  }, {
    key: "asObject",
    value: function asObject() {
      return {
        x: this.x,
        y: this.y
      };
    }
  }, {
    key: "asArray",
    value: function asArray() {
      return [this.x, this.y];
    }
  }, {
    key: "equals",
    value: function equals(vector) {
      return this === vector || this.x === vector.x && this.y === vector.y;
    }
  }, {
    key: "length",
    get: function get() {
      return Math.sqrt(this.x * this.x + this.y * this.y);
    }
  }]);

  return Vector;
})();

exports["default"] = Vector;
module.exports = exports["default"];
